index.d.ts

```diff
@@ -92,9 +97,12 @@ export function combineReducers<S>(reducers: ReducersMapObject<S>): Reducer<S>;
  * function to handle async actions in addition to actions. Middleware may
  * transform, delay, ignore, or otherwise interpret actions or async actions
  * before passing them to the next middleware.
+ *
+ * @template S unused, here only for backwards compatibility.
```

Comment:
I think we should put more thought into this. These changes are already breaking, so why should we bother about compatibility here?

The `S` type parameter brought much more misconception and inconvenience than benefit. It was initially added to allow us to strongly type-check dispatching thunks: as you can see [here](https://github.com/gaearon/redux-thunk/blob/v2.2.0/index.d.ts), we augment the `Dispatch` interface to add a new signature for thunks that uses `S`.

But in real-world it appeared that we rarely use `dispatch` as a method of the Store, where it has a fixed state type, but rather as an injected function (like in `react-redux`), where we have to declare its type manually:
```ts
const mapDispatchToProps = (dispatch: Dispatch<MyState>) => ...
```
For users who don't use `redux-thunk` this led to annoying need to write `Dispatch<any>` instead of just `Dispatch`.

We could just remove `S` altogether to simplify things; `redux-thunk` users would still be able to get partial type safety by declaring their state type in `ThunkAction` type parameters. The one major drawback of this is that it would make it much harder to migrate to the new typings.

Otherwise we should make this docstring point to the use case of `S`.

Reply:
As I said I'm happy to remove the state parameter and I definitely think this should be done at some point... If you think this PR is the right place for it I'll remove it.